File: android/src/main/java/com/margelo/nitro/wifiaware/support/ListenerStore.kt
package com.margelo.nitro.wifiaware.support

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.wifiaware.ListenerSubscription
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * A thread-safe registry of JS listeners for one event.
 *
 * Nitro callbacks are safe to invoke from any thread, but the registry is written from the Wi-Fi
 * Aware and connectivity callback threads while being read during emission, so it is backed by a
 * concurrent map. `remove()` is idempotent, matching the documented `ListenerSubscription`
 * contract.
 */
@Keep
@DoNotStrip
class ListenerStore<Payload> {
  private val listeners = ConcurrentHashMap<Long, (Payload) -> Unit>()
  private val nextToken = AtomicLong(0)

  /** Registers a listener and returns the handle JS uses to unregister it. */
  fun add(listener: (Payload) -> Unit): ListenerSubscription {
    val token = nextToken.getAndIncrement()
    listeners[token] = listener
    return ListenerSubscription(remove = { listeners.remove(token) })
  }

  /** Invokes every registered listener with [payload]. */
  fun emit(payload: Payload) {
    for (listener in listeners.values) {
      listener(payload)
    }
  }

  /** Whether anything is listening. Lets callers skip building an expensive payload. */
  val isEmpty: Boolean
    get() = listeners.isEmpty()

  /** Drops every listener, used when the owning object is torn down. */
  fun removeAll() {
    listeners.clear()
  }
}

File: ios/Support/ListenerStore.swift
import Foundation
import NitroModules

/// A thread-safe registry of JS listeners for one event.
///
/// Nitro callbacks are safe to invoke from any thread, but the registry itself is mutated from the
/// networking queues and read while emitting, so every access is serialised behind a lock.
///
/// `remove()` is idempotent, and a listener removed while an emission is already in flight may
/// still receive that one event — the contract documented on `ListenerSubscription`.
final class ListenerStore<Payload>: @unchecked Sendable {
  private let lock = NSLock()
  private var listeners: [Int64: (Payload) -> Void] = [:]
  private var nextToken: Int64 = 0

  /// Registers a listener and returns the handle JS uses to unregister it.
  func add(_ listener: @escaping (Payload) -> Void) -> ListenerSubscription {
    lock.lock()
    let token = nextToken
    nextToken += 1
    listeners[token] = listener
    lock.unlock()

    return ListenerSubscription(remove: { [weak self] in
      guard let self else { return }
      self.lock.lock()
      self.listeners.removeValue(forKey: token)
      self.lock.unlock()
    })
  }

  /// Invokes every registered listener with `payload`.
  ///
  /// The snapshot is taken under the lock and the callbacks run outside it, so a listener that
  /// unsubscribes (or subscribes) from inside its own callback cannot deadlock.
  func emit(_ payload: Payload) {
    lock.lock()
    let snapshot = Array(listeners.values)
    lock.unlock()

    for listener in snapshot {
      listener(payload)
    }
  }

  /// Whether anything is currently listening. Lets callers skip building an expensive payload.
  var isEmpty: Bool {
    lock.lock()
    defer { lock.unlock() }
    return listeners.isEmpty
  }

  /// Drops every listener, used when the owning object is torn down.
  func removeAll() {
    lock.lock()
    listeners.removeAll()
    lock.unlock()
  }
}
